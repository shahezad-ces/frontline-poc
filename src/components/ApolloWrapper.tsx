"use client";
import { ApolloProvider } from "@apollo/client/react";
import client from "@frontline/libs/apolloClient";

export default function ApolloWrapper({ children }: { children: React.ReactNode }) {
  return <ApolloProvider client={client}>{children}</ApolloProvider>;
}