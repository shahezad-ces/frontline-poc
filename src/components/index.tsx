import Categories from "./Categories/Categories";
import PaginationWrapper from "./PaginationWrapper/PaginationWrapper";
import ProductCard from "./ProductCard/ProductCard";
import ProductsFilters from "./ProductsFilters/ProductsFilters";
import AddToCartButton from "./AddToCartButton/AddToCartButton";
import Header from "./Header/Header";

export {
  Categories,
  PaginationWrapper,
  ProductCard,
  ProductsFilters,
  AddToCartButton,
  Header,
};
