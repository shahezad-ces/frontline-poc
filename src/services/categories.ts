import {
  GET_CATEGORIES_QUERY,
  GET_CATEGORY_DETAILS_QUERY,
} from "@frontline/graphql";
import { query } from "@frontline/libs/ApolloClient";
import { Category } from "@frontline/types";

export const getCategories = async (): Promise<Category[] | undefined> => {
  try {
    const { data } = await query<{ categories: Category[] }>({
      query: GET_CATEGORIES_QUERY,
      fetchPolicy: "cache-first",
    });

    return data?.categories;
  } catch (error) {
    console.error("GraphQL Error in getCategories:", error);
    return undefined;
  }
};

export const getCategoryDetails = async (
  id: number
): Promise<Category | undefined> => {
  if (!id) return undefined;
  try {
    const { data } = await query<{ category: Category }>({
      query: GET_CATEGORY_DETAILS_QUERY,
      variables: { id },
      fetchPolicy: "cache-first",
    });

    return data?.category;
  } catch (error) {
    console.error("GraphQL Error in getCategoryDetails:", error);
    return undefined;
  }
};
