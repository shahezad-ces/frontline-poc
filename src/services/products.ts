import {
  GET_PRODUCTS_QUERY,
  GET_PRODUCT_DETAILS_QUERY,
} from "@frontline/graphql";
import { query } from "@frontline/libs/ApolloClient";
import type { Product } from "@frontline/types";

export interface GetProductsParams {
  categoryId?: number;
  price?: number;
  price_min?: number;
  price_max?: number;
  offset?: number;
  limit?: number;
}

export const getProducts = async (
  params: GetProductsParams
): Promise<Product[] | undefined> => {
  const {
    categoryId,
    price,
    price_min,
    price_max,
    offset = 0,
    limit = 20,
  } = params;
  const variables = {
    offset,
    limit,
    ...(categoryId !== undefined && { categoryId }),
    ...(price !== undefined && { price }),
    ...(price_min !== undefined && { price_min }),
    ...(price_max !== undefined && { price_max }),
  };

  const { data } = await query<{ products: Product[] }>({
    query: GET_PRODUCTS_QUERY,
    variables,
    fetchPolicy: "cache-first",
  });

  return data?.products;
};

export const getProductDetails = async (
  productId: number
): Promise<Product | undefined> => {
  try {
    const { data } = await query<{ product: Product }>({
      query: GET_PRODUCT_DETAILS_QUERY,
      variables: { id: productId },
      fetchPolicy: "cache-first",
    });

    return data?.product;
  } catch (error) {
    console.error("GraphQL Error:", error);
    return undefined; // fallback
  }
};
