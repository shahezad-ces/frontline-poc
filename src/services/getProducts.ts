import { GET_PRODUCTS_QUERY } from "@frontline/graphql/products";
import { fetchGraphQL } from "@frontline/libs/apolloClient";
import { Product } from "@frontline/types/product";

export async function getProducts(): Promise<Product[]> {
  const variables = {
    offset: 0,
    limit: 20,
  };

  const data = await fetchGraphQL<{ products: Product[] }>({
    query: GET_PRODUCTS_QUERY,
    variables,
  });

  return data.products;
}
