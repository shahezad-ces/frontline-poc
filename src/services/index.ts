import { getCategories, getCategoryDetails } from "./categories";
import { getProducts, getProductDetails } from "./products";

export { getCategories, getCategoryDetails, getProducts, getProductDetails };
