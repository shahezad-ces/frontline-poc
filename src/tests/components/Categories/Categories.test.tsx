import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Categories } from "@frontline/components";
import { getCategories } from "@frontline/services";
import type { Category } from "@frontline/types";

jest.mock("@frontline/services", () => ({
  getCategories: jest.fn(),
}));

const mockGetCategories = getCategories as jest.Mock;

const mockCategories: Category[] = [
  { id: 1, name: "Electronics", image: "test" },
  { id: 2, name: "Clothing", image: "test" },
  { id: 3, name: "Books", image: "test" },
];

describe("Categories Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders categories list when API returns data", async () => {
    mockGetCategories.mockResolvedValue(mockCategories);

    render(await Categories({ selectedCategory: null }));

    expect(screen.getByText(/Categories/i)).toBeInTheDocument();
    expect(screen.getByText(/Electronics/i)).toBeInTheDocument();
    expect(screen.getByText(/Clothing/i)).toBeInTheDocument();
    expect(screen.getByText(/Books/i)).toBeInTheDocument();
  });

  it("handles empty categories array", async () => {
    mockGetCategories.mockResolvedValue([]);

    render(await Categories({ selectedCategory: null }));

    expect(screen.getByText(/Categories/i)).toBeInTheDocument();
    expect(screen.queryByRole("link")).not.toBeInTheDocument();
  });

  it("handles null or undefined response from API", async () => {
    mockGetCategories.mockResolvedValue(null);

    render(await Categories({ selectedCategory: null }));

    expect(screen.getByText(/Categories/i)).toBeInTheDocument();
    expect(screen.queryByRole("link")).not.toBeInTheDocument();
  });

  it("matches snapshot", async () => {
    mockGetCategories.mockResolvedValue(mockCategories);

    const { asFragment } = render(await Categories({ selectedCategory: null }));
    expect(asFragment()).toMatchSnapshot();
  });
});
