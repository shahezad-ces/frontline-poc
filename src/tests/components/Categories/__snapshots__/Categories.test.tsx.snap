// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Categories Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="shadow border border-gray-200 rounded-lg p-4 bg-white"
  >
    <h2
      class="text-2xl font-semibold mb-2"
    >
      Categories
    </h2>
    <ul
      class="space-y-1"
    >
      <li>
        <a
          href="/products?categoryId=1"
        >
          Electronics
        </a>
      </li>
      <li>
        <a
          href="/products?categoryId=2"
        >
          Clothing
        </a>
      </li>
      <li>
        <a
          href="/products?categoryId=3"
        >
          Books
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
